Test tutorials (#1587)
denoland/docs: examples/tutorials/mocking.md

---
title: "Testing in isolation with mocks"

> description: "Master the art of mocking in your unit tests in with this comprehensive guide. Learn how spies, stubs, and fake time, and other Deno helpers let you to isolate your code during testing, making your unit tests more reliable and efficient."

Reviewer: While this is a good intro / teaser, I think this is too long for a meta description. Sadly these need to be pretty short in order to perform well. (and work in our OG images)

It gives me the ick a little, but tools like this can sometime help with crafting titles and meta descriptions: https://rankmath.com/tools/meta-tag-analyzer/

FWIW I muddle around in it for a few minutes and came back with this:

```suggestion
description: "Master the art of mocking in your unit tests. Learn how spies, stubs, fake time, and other Deno tools let you improve your code and confidence"
```

want to avoid during testing.

We set up the function under test, in this case the `hasAdminAccess()` function.
We want to test whether it:

- Calls the `getCurrentUser()` function to get a user object
- Checks if the user's role is "admin"
- Returns a boolean indicating whether the user has admin access

Next we create a test named `hasAdminAccess with a stubbed user` and set up a
stub for the `getCurrentUser` function. This will replace the real
implementation with one that returns a user with a `guest` role.

We run the test with the stubbed function, it will call `hasAdminAccess` with a
user ID. Even though the real function would return a user with `admin` role,
our stub returns `guest`, so we can assert that `hasAdminAccess` returns `false`
(since our stub returns a non-admin user).

We can change the stub behavior to return `admin` instead and assert that the
function now returns `true`.

At the end we use a `finally` block to ensure the original function is restored
so that we don't accidentally affect other tests.

### Stubbing environment variables

For deterministic testing, you often need to control environment variables.

> Deno's Standard Library provides utilities to achieve this:

Reviewer: both this code block, and the code clock below are quite long. I didn't spot that on each occasion you unpacked and explained them in detail underneath, so I spent time parsing them myself before discovering that helpful description underneath.

I suggest leaving a small hint that this is coming up so that the reader knows that they can jump back and forth between the code and your breakdown of it. Perhaps something like:

```suggestion
Deno's Standard Library provides utilities to achieve this as we see in this code, and the explanation of it which follows below
```

---
title: "Testing in isolation with mocks"
description: "Master the art of mocking in your unit tests. Learn how spies, stubs, fake time, and other Deno tools let you improve your code and confidence"
url: /examples/mocking_tutorial/
---

# Testing in isolation with mocks

This guide builds on the
[basics of testing in Deno](/examples/testing_tutorial/) to focus specifically
on mocking techniques that help you isolate your code during testing.

## Table of Contents

- [Introduction](#introduction)
- [Spying](#spying)
- [Stubbing](#stubbing)
- [Faking time](#faking-time)
- [Advanced mocking patterns](#advanced-mocking-patterns)
- [Real-world example](#real-world-example)
- [Summary](#summary)

## Introduction

For effective unit testing, you'll often need to mock dependencies. Mocking is a
technique used in testing where you replace real dependencies with simulated
versions that you can control. This is particularly useful when testing

> components that interact with external services, such as APIs or databases.

Reviewer 2: Nit: I'm worried about the wording of `mock dependencies` here because that's commonly associated with mocking full modules or npm libraries. That's something we don't support in Deno's test runner.

Deno provides powerful mocking utilities through the standard library without

> requiring additional frameworks, making your tests more reliable and faster.

Reviewer 2: Q: how does using mocks make my "tests faster"?

### Spying

In Deno, you can [`spy`](https://jsr.io/@std/testing/doc/mock#spying) on a
function to test that it behaves as expected.

By using spies, you can verify that your code interacts correctly with its

> dependencies without setting up complex infrastructure.

Reviewer 2: Nit: I feel like test spies is less about testing if a function behaves as expected and more about being able to create dummy functions on the fly to assert that it was called with the right arguments and stuff.

Author: have updated
